### PR TITLE
 Add TLS support for client-bricks communication

### LINKED ISSUE #214: TLS encryption for client to server (brick communication)
GD2 should have TLS/SSL or similar encryption support implemented.

### LINKED ISSUE #971: Add support for glusterfs client authorization
GlusterFS supports many methods to allow or disallow certain clients to access volume. Users can specify patterns of host/IP addresses using options such as `auth.allow` and `auth.reject`

### PR DESCRIPTION
**Pull request by prashanthpai**
This change enables setting up TLS for client to bricks communication.

Unlike glusterd1 which requires two separate options - `client.ssl` and
`server.ssl` to be turned on, turning TLS on for a volume in glusterd2
is a single volume set operation. TLS can be enabled for a volume as
follows:

    $ glustercli volume set <volname> --advanced tls on

To achieve this, when xlator shared ojects are loaded during startup,
options present in transport layer (socket.so and rdma.so) are injected
into list of options loaded from protocol layer (server.so and
client.so). The client and brick volfile expects the transport options
to be in client and server xlator sections.

This PR also tests that client authorization works as expected which can be used as follows:

```sh
$ glustercli volume set test --advanced server.ssl-allow <client-CN-name>
```
This is dependent on: https://review.gluster.org/#/c/20472

Other fixes:
* Add support for xlator options containing dots
* volume-set: Validate expanded group options (bug fix)
* Set default boolean value properly for group option
* Fix bug in validating internet addresses

Closes #214 
Closes #971

### CODE REVIEW

**Comment by prashanthpai** [comment]
Tested as follows:

**Setting up certs**

```sh
# openssl genrsa -out /etc/ssl/glusterfs.key 2048
# openssl req -new -x509 -key /etc/ssl/glusterfs.key -subj /CN=Anyone -out /etc/ssl/glusterfs.pem
# cp /etc/ssl/glusterfs.pem /etc/ssl/glusterfs.ca
```

**Enabling TLS for a volume**

```sh
glustercli volume set test --advanced tls on
```

```sh
$ glustercli volume info test

Volume Name: test
Type: Distribute
Volume ID: 12eb8aa0-8c2a-4aeb-b957-aba2d74f58b3
State: Created
Transport-type: tcp
Options:
    client.transport.socket.ssl-enabled: on
    server.transport.socket.ssl-enabled: on
Number of Bricks: 1
Brick1: 127.0.0.1:/export/brick1/data
```

This shall be added to `e2e` tests.

File: glusterd2/commands/volumes/volume-option.go [hunk 1]
```diff
@@ -43,7 +43,7 @@ func optionSetValidate(c transaction.TxnCtx) error {
 		return err
 	}
 
-	if err := validateXlatorOptions(req.Options, &volinfo); err != nil {
+	if err := validateXlatorOptions(options, &volinfo); err != nil {
 		return fmt.Errorf("Validation failed for volume option:: %s", err.Error())
```
Comment (rishubhjain) [inline glusterd2/commands/volumes/volume-option.go]: TRIVIAL :  @prashanthpai @Madhu-1 Should this statement start with lower case as well? Not part of changes made in this PR though.

Comment (prashanthpai) [reply glusterd2/commands/volumes/volume-option.go]: Sure

**Comment by prashanthpai** [comment]
retest this please

**Comment by prashanthpai** [comment]
This implements the authentication of clients. Authorisation of clients using `auth.ssl-allow` is yet to implemented.

**Comment by prashanthpai** [comment]
retest this please

**Comment by prashanthpai** [comment]
retest this please

File: glusterd2/xlator/options/utils.go [hunk 2]
```diff
@@ -3,26 +3,51 @@ package options
...
+
+	if utils.StringInSlice(tmp[0], validGraphs[:]) {
+		// valid graph present
+		if len(tmp) < 3 {
+			// must be of the form <graph>.<xlator>.<name>
+			return graph, xlator, optName, InvalidKeyError(k)
+		}
+		graph = tmp[0]
+		xlator = tmp[1]
+		optName = k[len(graph)+1+len(xlator)+1:]
```
[the remark this answers is not in the document]
Comment (prashanthpai) [reply glusterd2/xlator/options/utils.go]: Done

**Comment by prashanthpai** [comment]
retest this please

File: glusterd2/commands/volumes/grouped-options.go [hunk 3]
```diff
@@ -129,6 +129,12 @@ var defaultGroupOptions = map[string]*api.OptionGroup{
 			{Name: "features.shard", OnValue: "on"},
 			{Name: "user.cifs", OnValue: "off"}},
 		"Enable this profile, if the Gluster Volume is used to store virtual machines"},
+	"tls": {"tls",
+		[]api.VolumeOption{
+			{Name: "server.transport.socket.ssl-enabled", OnValue: "on"},
+			{Name: "client.transport.socket.ssl-enabled", OnValue: "on"},
```
Comment (kshlm) [inline glusterd2/commands/volumes/grouped-options.go]: We should possibly try to get the `socket.ssl-*` options to `socket.tls-*`. Makes this more the option more consistent.

**Pull request merged**